wcfsetup/install/files/lib/action/FileDownloadAction.class.php

```diff
@@ -0,0 +1,82 @@
+<?php
+
+namespace wcf\action;
+
+use Laminas\Diactoros\Response;
+use Laminas\Diactoros\Stream;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use Psr\Http\Server\RequestHandlerInterface;
+use wcf\data\file\File;
+use wcf\http\Helper;
+use wcf\system\exception\IllegalLinkException;
+use wcf\system\exception\PermissionDeniedException;
+use wcf\util\FileUtil;
+
+/**
+ * Offers a file for download.
+ *
+ * @author Alexander Ebert
+ * @copyright 2001-2024 WoltLab GmbH
+ * @license GNU Lesser General Public License <http://opensource.org/licenses/lgpl-license.php>
+ * @since 6.1
+ */
+final class FileDownloadAction implements RequestHandlerInterface
+{
+    public function handle(ServerRequestInterface $request): ResponseInterface
+    {
+        $parameters = Helper::mapQueryParameters(
+            $request->getQueryParams(),
+            <<<'EOT'
+                array {
+                    id: positive-int
+                }
+                EOT,
+        );
+
+        $file = new File($parameters['id']);
+        if (!$file->fileID) {
+            throw new IllegalLinkException();
+        }
+
+        $processor = $file->getProcessor();
+        if ($processor === null) {
+            throw new IllegalLinkException();
+        }
+
+        if (!$processor->canDownload($file)) {
+            throw new PermissionDeniedException();
+        }
+
+        $filename = $file->getPathname();
+        $response = new Response(
+            new Stream($filename),
+        );
+
+        $mimeType = FileUtil::getMimeType($filename);
+
+        // TODO: This should use `FileReader` instead.
+
```

Comment:
Should it? The PSR-7/-15 implementation should already support streaming: That's the whole purpose of the `$response->getBody()` being a stream.

Reply:
I’m not sure, the `FileReader` supports all sorts of different headers and range handling, this is a very early comment too. I will look into it, thanks.

Reply:
Header handling is much cleaner and easier with plain PSR-7. As for the Range header support, I recommend:

1. Create a [helper function](https://github.com/WoltLab/WCF/blob/master/wcfsetup/install/files/lib/http/Helper.class.php) to parse the header, returning $offset + $length.
2. Possibly perform some validation of the returned offset + length against the filesize.
3. Set the appropriate response headers.
4. Wrap the body Stream into [Guzzle's LimitStream](https://github.com/guzzle/psr7/blob/2.6/src/LimitStream.php).

In fact all 4 steps could be wrapped into a single function that takes a `ResponseInterface` and returns an updated `ResponseInterface` with headers applied and body wrapped.